静态图高阶微分计算结果不符合预期

### bug描述 Describe the Bug

``` py
import paddle
from paddle import nn


class Model(nn.Layer):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(2, 1)

    def forward(self, x, y):
        z = paddle.concat([x, y], axis=-1)
        z = self.fc1(z)
        z = nn.functional.silu(z)

        gx = paddle.grad(z, x, create_graph=True)[0]
        return paddle.abs(gx).pow(2).sum()

model = Model()

# data
N = 2
x = paddle.randn([N, 1])
x.stop_gradient = False
y = paddle.randn([N, 1])
y.stop_gradient = False

# dynamic
loss_dynamic = model(x, y)
loss_dynamic.backward()

grads1 = [param.grad.clone().detach() for param in model.parameters()]

model.clear_gradients()

# static
st_model = paddle.jit.to_static(model)

loss_static = st_model(x, y)
loss_static.backward()

grads2 = [param.grad.clone().detach() for param in model.parameters()]

# check
import numpy as np
for p1, p2 in zip(grads1, grads2):
    np.testing.assert_allclose(p1.numpy(), p2.numpy(), 1e-3, 1e-3)
```

运行命令：

``` sh
FLAGS_prim_all=False FLAGS_enable_pir_in_executor=true FLAGS_enable_pir_api=True python test_run.py
```

结果：

``` sh
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001

Mismatched elements: 8 / 8 (100%)
Max absolute difference: 0.08932383
Max relative difference: 3.0927124
 x: array([[ 0.050079,  0.003131,  0.12114 , -0.012295],
       [-0.106463,  0.056415, -0.26742 ,  0.029972]], dtype=float32)
 y: array([[ 0.012236, -0.017195,  0.031816, -0.003915],
       [-0.068708,  0.077533, -0.178814,  0.021687]], dtype=float32)
```

由于silu没有二阶大算子，因此运行时预期应该报错，而不是仍然进行计算并给出错误结果；

如果把激活函数从silu换成配二阶算子的tanh，则可以通过测试。

### 其他补充信息 Additional Supplementary Information

_No response_

--- a/python/paddle/autograd/backward_utils.py
+++ b/python/paddle/autograd/backward_utils.py
@@ -25,6 +25,73 @@ from paddle.base.libpaddle.pir import (
 )
 from paddle.base.wrapped_decorator import signature_safe_contextmanager
 
+# TODO: Consider a better way to mark these ops has no grad op.
+# Such as use a new trait to mark these ops.
+ALLOW_NO_GRAD_OPS = [
+    # Compare ops
+    "pd_op.equal",
+    "pd_op.equal_",
+    "pd_op.not_equal",
+    "pd_op.not_equal_",
+    "pd_op.less_than",
+    "pd_op.less_than_",
+    "pd_op.less_equal",
+    "pd_op.less_equal_",
+    "pd_op.greater_than",
+    "pd_op.greater_than_",
+    "pd_op.greater_equal",
+    "pd_op.greater_equal_",
+    # Logical ops
+    "pd_op.logical_and",
+    "pd_op.logical_and_",
+    "pd_op.logical_not",
+    "pd_op.logical_not_",
+    "pd_op.logical_or",
+    "pd_op.logical_or_",
+    "pd_op.logical_xor",
+    "pd_op.logical_xor_",
+    # Bitwise ops
+    "pd_op.bitwise_and",
+    "pd_op.bitwise_and_",
+    "pd_op.bitwise_left_shift",
+    "pd_op.bitwise_left_shift_",
+    "pd_op.bitwise_not",
+    "pd_op.bitwise_not_",
+    "pd_op.bitwise_or",
+    "pd_op.bitwise_or_",
+    "pd_op.bitwise_right_shift",
+    "pd_op.bitwise_right_shift_",
+    "pd_op.bitwise_xor",
+    "pd_op.bitwise_xor_",
+    # Array ops
+    "pd_op.assign_array",
+    "pd_op.array_length",
+    "pd_op.slice_array",
+    "pd_op.slice_array_dense",
+    "pd_op.assign_array",
+    "pd_op.assign_array_",
+    "pd_op.create_array",
+    "pd_op.create_array_like",
+    "pd_op.array_read",
+    "pd_op.array_write_",
+    "pd_op.array_pop",
+    # Others
+    "pd_op.remainder",
+    "pd_op.argmax",
+    "pd_op.print",
+    "pd_op.accuracy",
+    "pd_op.uniform",
+    "pd_op.gaussian",
+    "pd_op.bernoulli",
+    "pd_op.full_like",
+    "pd_op.assign_value_",
+    "pd_op.nextafter",
+    "pd_op.isnan",
+    "pd_op.isinf",
+    "pd_op.all",
+    "pd_op.any",
+]
+
 
 class ValueWrapper:
     def __init__(self, value) -> None:
@@ -279,6 +346,11 @@ def some_in_set(value_list, value_set):
 
 def is_control_flow(op):
     return op.name() == "pd_op.if" or op.name() == "pd_op.while"
+
+
+def is_builtin_op(op):
+    dialect_name, opname = op.name().split(".")
+    return dialect_name == "builtin"
 
 
 def update_no_grad_set_by_stopgradient(block, no_grad_set):

--- a/python/paddle/autograd/ir_backward.py
+++ b/python/paddle/autograd/ir_backward.py
@@ -838,7 +838,7 @@ def append_backward_ops(
                 else:
                     if op.name() not in ALLOW_NO_GRAD_OPS:
                         raise ValueError(
-                            f"op {op.name()} has no grad op, consider enable prim to decompose it."
+                            f"op '{op.name()}' has no grad op, consider enable prim to decompose it."
                         )
                     state.op_to_opgrad[op] = []
 

--- a/python/paddle/autograd/ir_backward.py
+++ b/python/paddle/autograd/ir_backward.py
@@ -55,6 +55,8 @@ from paddle.base.libpaddle.pir import (
 """
 __all__ = ['grad', 'calc_gradient', 'calc_gradient_helper']
 
+ALLOW_NO_GRAD_OPS = ["pd_op.full_like"]
+
 
 def append_full_like(float_value, copy_value, value, state, backward_ops):
     with paddle.amp.auto_cast(enable=False):
@@ -834,7 +836,10 @@ def append_backward_ops(
                         else:
                             state.op_to_opgrad[op] = []
                 else:
-                    logging.warning("%s op has no grad op", op.name())
+                    if op.name() not in ALLOW_NO_GRAD_OPS:
+                        raise ValueError(
+                            f"op {op.name()} has no grad op, consider enable prim to decompose it."
+                        )
                     state.op_to_opgrad[op] = []
 
         if fwd_block != bwd_block:
@@ -1202,9 +1207,11 @@ def append_backward(loss, parameter_list=None, no_grad_set=None):
         input_inputs_grad.append(
             (
                 input,
-                input_to_inputgrad_map[input][0][0]
-                if input_to_inputgrad_map[input] != []
-                else None,
+                (
+                    input_to_inputgrad_map[input][0][0]
+                    if input_to_inputgrad_map[input] != []
+                    else None
+                ),
             )
         )
 

--- a/python/paddle/autograd/ir_backward.py
+++ b/python/paddle/autograd/ir_backward.py
@@ -80,6 +80,19 @@ ALLOW_NO_GRAD_OPS = [
     "pd_op.logical_or_",
     "pd_op.logical_xor",
     "pd_op.logical_xor_",
+    # Bitwise ops
+    "pd_op.bitwise_and",
+    "pd_op.bitwise_and_",
+    "pd_op.bitwise_left_shift",
+    "pd_op.bitwise_left_shift_",
+    "pd_op.bitwise_not",
+    "pd_op.bitwise_not_",
+    "pd_op.bitwise_or",
+    "pd_op.bitwise_or_",
+    "pd_op.bitwise_right_shift",
+    "pd_op.bitwise_right_shift_",
+    "pd_op.bitwise_xor",
+    "pd_op.bitwise_xor_",
     # Array ops
     "pd_op.assign_array",
     "pd_op.array_length",
@@ -105,6 +118,8 @@ ALLOW_NO_GRAD_OPS = [
     "pd_op.nextafter",
     "pd_op.isnan",
     "pd_op.isinf",
+    "pd_op.all",
+    "pd_op.any",
 ]
 
 

--- a/python/paddle/autograd/ir_backward.py
+++ b/python/paddle/autograd/ir_backward.py
@@ -18,6 +18,7 @@ import logging
 
 import paddle.pir
 from paddle.autograd.backward_utils import (
+    ALLOW_NO_GRAD_OPS,
     State,
     ValueDict,
     ValueSet,
@@ -32,6 +33,7 @@ from paddle.autograd.backward_utils import (
     get_real_op_inputs,
     get_split_op,
     inverse_sort_op,
+    is_builtin_op,
     is_control_flow,
     is_inplace_net,
     parent_total_ops,
@@ -54,78 +56,6 @@ from paddle.base.libpaddle.pir import (
     calc_gradient_helper: for dygraph to static .
 """
 __all__ = ['grad', 'calc_gradient', 'calc_gradient_helper']
-
-# TODO: Consider a better way to mark these ops has no grad op.
-# Such as use a new trait to mark these ops.
-ALLOW_NO_GRAD_OPS = [
-    # Compare ops
-    "pd_op.equal",
-    "pd_op.equal_",
-    "pd_op.not_equal",
-    "pd_op.not_equal_",
-    "pd_op.less_than",
-    "pd_op.less_than_",
-    "pd_op.less_equal",
-    "pd_op.less_equal_",
-    "pd_op.greater_than",
-    "pd_op.greater_than_",
-    "pd_op.greater_equal",
-    "pd_op.greater_equal_",
-    # Logical ops
-    "pd_op.logical_and",
-    "pd_op.logical_and_",
-    "pd_op.logical_not",
-    "pd_op.logical_not_",
-    "pd_op.logical_or",
-    "pd_op.logical_or_",
-    "pd_op.logical_xor",
-    "pd_op.logical_xor_",
-    # Bitwise ops
-    "pd_op.bitwise_and",
-    "pd_op.bitwise_and_",
-    "pd_op.bitwise_left_shift",
-    "pd_op.bitwise_left_shift_",
-    "pd_op.bitwise_not",
-    "pd_op.bitwise_not_",
-    "pd_op.bitwise_or",
-    "pd_op.bitwise_or_",
-    "pd_op.bitwise_right_shift",
-    "pd_op.bitwise_right_shift_",
-    "pd_op.bitwise_xor",
-    "pd_op.bitwise_xor_",
-    # Array ops
-    "pd_op.assign_array",
-    "pd_op.array_length",
-    "pd_op.slice_array",
-    "pd_op.slice_array_dense",
-    "pd_op.assign_array",
-    "pd_op.assign_array_",
-    "pd_op.create_array",
-    "pd_op.create_array_like",
-    "pd_op.array_read",
-    "pd_op.array_write_",
-    "pd_op.array_pop",
-    # Others
-    "pd_op.remainder",
-    "pd_op.argmax",
-    "pd_op.print",
-    "pd_op.accuracy",
-    "pd_op.uniform",
-    "pd_op.gaussian",
-    "pd_op.bernoulli",
-    "pd_op.full_like",
-    "pd_op.assign_value_",
-    "pd_op.nextafter",
-    "pd_op.isnan",
-    "pd_op.isinf",
-    "pd_op.all",
-    "pd_op.any",
-]
-
-
-def is_builtin_op(op):
-    dialect_name, opname = op.name().split(".")
-    return dialect_name == "builtin"
 
 
 def append_full_like(float_value, copy_value, value, state, backward_ops):

--- a/python/paddle/autograd/ir_backward.py
+++ b/python/paddle/autograd/ir_backward.py
@@ -55,7 +55,62 @@ from paddle.base.libpaddle.pir import (
 """
 __all__ = ['grad', 'calc_gradient', 'calc_gradient_helper']
 
-ALLOW_NO_GRAD_OPS = ["pd_op.full_like"]
+# TODO: Consider a better way to mark these ops has no grad op.
+# Such as use a new trait to mark these ops.
+ALLOW_NO_GRAD_OPS = [
+    # Compare ops
+    "pd_op.equal",
+    "pd_op.equal_",
+    "pd_op.not_equal",
+    "pd_op.not_equal_",
+    "pd_op.less_than",
+    "pd_op.less_than_",
+    "pd_op.less_equal",
+    "pd_op.less_equal_",
+    "pd_op.greater_than",
+    "pd_op.greater_than_",
+    "pd_op.greater_equal",
+    "pd_op.greater_equal_",
+    # Logical ops
+    "pd_op.logical_and",
+    "pd_op.logical_and_",
+    "pd_op.logical_not",
+    "pd_op.logical_not_",
+    "pd_op.logical_or",
+    "pd_op.logical_or_",
+    "pd_op.logical_xor",
+    "pd_op.logical_xor_",
+    # Array ops
+    "pd_op.assign_array",
+    "pd_op.array_length",
+    "pd_op.slice_array",
+    "pd_op.slice_array_dense",
+    "pd_op.assign_array",
+    "pd_op.assign_array_",
+    "pd_op.create_array",
+    "pd_op.create_array_like",
+    "pd_op.array_read",
+    "pd_op.array_write_",
+    "pd_op.array_pop",
+    # Others
+    "pd_op.remainder",
+    "pd_op.argmax",
+    "pd_op.print",
+    "pd_op.accuracy",
+    "pd_op.uniform",
+    "pd_op.gaussian",
+    "pd_op.bernoulli",
+    "pd_op.full_like",
+    "pd_op.assign_value_",
+    "pd_op.nextafter",
+    "pd_op.isnan",
+    "pd_op.isinf",
+]
+
+
+def is_builtin_op(op):
+    dialect_name, opname = op.name().split(".")
+    return dialect_name == "builtin"
 
 
 def append_full_like(float_value, copy_value, value, state, backward_ops):
@@ -836,7 +891,10 @@ def append_backward_ops(
                         else:
                             state.op_to_opgrad[op] = []
                 else:
-                    if op.name() not in ALLOW_NO_GRAD_OPS:
+                    if (
+                        not is_builtin_op(op)
+                        and op.name() not in ALLOW_NO_GRAD_OPS
+                    ):
                         raise ValueError(
                             f"op '{op.name()}' has no grad op, consider enable prim to decompose it."
                         )

--- a/test/dygraph_to_static/test_high_order_net.py
+++ b/test/dygraph_to_static/test_high_order_net.py
@@ -1,0 +1,58 @@
+# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from dygraph_to_static_utils import (
+    Dy2StTestBase,
+    test_ast_only,
+    test_pir_only,
+)
+
+import paddle
+
+
+class HighOrderNet(paddle.nn.Layer):
+    def __init__(self):
+        super().__init__()
+        self.linear = paddle.nn.Linear(3, 4, bias_attr=False)
+
+    def forward(self, x):
+        y = self.linear(x)
+        z = paddle.pow(y, 2)
+        x_grad = paddle.grad(z, x, create_graph=True)[0]
+        x_grad_grad = paddle.grad(x_grad, x, create_graph=True)[0]
+        return x_grad_grad.mean()
+
+
+class TestBackwardHasNoGradError(Dy2StTestBase):
+    @test_ast_only
+    @test_pir_only
+    def test_backward_has_no_grad_error(self):
+        net = HighOrderNet()
+        static_net = paddle.jit.to_static(net, full_graph=True)
+
+        x = paddle.to_tensor([[1, 1, 1], [1, 1, 1]], 'float32')
+        x.stop_gradient = False
+
+        with self.assertRaisesRegex(
+            ValueError,
+            "op 'pd_op.matmul_double_grad' has no grad op, consider enable prim to decompose it.",
+        ):
+            x_grad_grad = static_net(x)
+            x_grad_grad.backward()
+
+
+if __name__ == "__main__":
+    unittest.main()

